In [1]:
import pandas as pd
import numpy as np
import datetime as dt
import warnings
import seaborn as sns
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
# read car assignment data
car = pd.read_csv("car-assignments.csv")
#make a full name
car['FullName'] = car[['FirstName', 'LastName']].apply(lambda x: ' '.join(x), axis=1)
#drop separate name cols
car.drop(columns=['FirstName', 'LastName'], inplace=True)

In [3]:
car.sort_values(by=['CurrentEmploymentType'])

CarID   CurrentEmploymentType        CurrentEmploymentTitle  \
13   14.0             Engineering     Engineering Group Manager   
25   26.0             Engineering            Drill Site Manager   
24   25.0             Engineering                     Geologist   
32   33.0             Engineering              Drill Technician   
18   19.0             Engineering          Hydraulic Technician   
17   18.0             Engineering                     Geologist   
26   27.0             Engineering              Drill Technician   
10   11.0             Engineering          Hydraulic Technician   
27   28.0             Engineering              Drill Technician   
1     2.0             Engineering                      Engineer   
2     3.0             Engineering                      Engineer   
6     7.0             Engineering              Drill Technician   
8     9.0             Engineering              Drill Technician   
9    10.0               Executive                       SVP/CIO   
30   31.0               Executive                 President/CEO   
34   35.0               Executive  Environmental Safety Advisor   
31   32.0               Executive                       SVP/COO   
3     4.0               Executive                       SVP/CFO   
28   29.0              Facilities      Facilities Group Manager   
42    NaN              Facilities                  Truck Driver   
41    NaN              Facilities                  Truck Driver   
35    NaN              Facilities                  Truck Driver   
36    NaN              Facilities                  Truck Driver   
37    NaN              Facilities                  Truck Driver   
38    NaN              Facilities                  Truck Driver   
39    NaN              Facilities                  Truck Driver   
40    NaN              Facilities                  Truck Driver   
43    NaN              Facilities                  Truck Driver   
0     1.0  Information Technology                   IT Helpdesk   
16   17.0  Information Technology                 IT Technician   
7     8.0  Information Technology                 IT Technician   
5     6.0  Information Technology              IT Group Manager   
4     5.0  Information Technology                 IT Technician   
29   30.0                Security        Security Group Manager   
22   23.0                Security                Badging Office   
20   21.0                Security             Perimeter Control   
19   20.0                Security              Building Control   
33   34.0                Security             Perimeter Control   
15   16.0                Security             Perimeter Control   
14   15.0                Security                  Site Control   
12   13.0                Security                  Site Control   
11   12.0                Security                  Site Control   
23   24.0                Security             Perimeter Control   
21   22.0                Security                Badging Office   

              FullName  
13       Lidelse Dedos  
25          Marin Onda  
24       Adra Nubarron  
32     Brand Tempestad  
18         Vira Frente  
17     Birgitta Frente  
26         Kare Orilla  
10        Gustav Cazar  
27         Elsa Orilla  
1           Lars Azada  
2          Felix Balas  
6      Isande Borrasca  
8          Axel Calzas  
9   Ada Campo-Corrente  
30   Sten Sanjorge Jr.  
34   Willem Vasco-Pais  
31         Orhan Strum  
3      Ingrid Barranco  
28       Bertrand Ovan  
42          Irene Nant  
41   Cecilia Morluniau  
35        Albina Hafon  
36      Benito Hawelon  
37     Claudio Hawelon  
38           Henk Mies  
39      Valeria Morlun  
40         Adan Morlun  
43      Dylan Scozzese  
0        Lucas Alcazar  
16         Sven Flecha  
7         Nils Calixto  
5        Linnea Bergen  
4            Isak Baza  
29       Felix Resumir  
22         Varja Lagos  
20      Hennie Osvaldo  
19        Stenig Fusil  
33         Edvard Vann  
15           

In [4]:
# read GPS data
gps = pd.read_csv("gps.csv")
gps["Timestamp"] = gps["Timestamp"].apply(lambda x: dt.datetime.strptime(x,"%m/%d/%Y %H:%M:%S"))
gps['date'] = gps['Timestamp'].dt.date
# The day of the week with Monday=0, Sunday=6
gps["day_of_week"] = gps['Timestamp'].dt.dayofweek
gps["weekend"] = gps['Timestamp'].dt.dayofweek > 4

# Converting hour
def hour_converter(x):
    if (x >= 6) and (x < 11):
        return 'Morning'
    elif (x >= 11) and (x < 16):
        return'Afternoon'
    elif (x >= 16) and (x < 20):
        return'Evening'
    elif (x >= 20) and (x < 24):
        return'Night'
    elif (x >= 0) and (x < 3):
        return'Midnight'
    elif (x >= 3) and (x < 6):
        return 'Early Morning'

gps['hour'] = pd.to_datetime(gps['Timestamp']).dt.hour
gps['time'] = gps['hour'].apply(hour_converter)


gps.head(10)


Timestamp  id        lat       long        date  day_of_week  \
0 2014-01-06 06:28:01  35  36.076225  24.874689  2014-01-06            0   
1 2014-01-06 06:28:01  35  36.076220  24.874596  2014-01-06            0   
2 2014-01-06 06:28:03  35  36.076211  24.874443  2014-01-06            0   
3 2014-01-06 06:28:05  35  36.076217  24.874253  2014-01-06            0   
4 2014-01-06 06:28:06  35  36.076214  24.874167  2014-01-06            0   
5 2014-01-06 06:28:07  35  36.076191  24.874056  2014-01-06            0   
6 2014-01-06 06:28:09  35  36.076194  24.873906  2014-01-06            0   
7 2014-01-06 06:28:10  35  36.076183  24.873814  2014-01-06            0   
8 2014-01-06 06:28:11  35  36.076167  24.873737  2014-01-06            0   
9 2014-01-06 06:28:12  35  36.076177  24.873621  2014-01-06            0   

   weekend  hour     time  
0    False     6  Morning  
1    False     6  Morning  
2    False     6  Morning  
3    False     6  Morning  
4    False     6  Morning  
5    False     6  Morning  
6    False     6  Morning  
7    False     6  Morning  
8    False     6  Morning  
9    False     6  Morning

In [5]:
# Include Truck Data
gps.to_csv(r'gps_fixed.csv')

In [6]:
np.sort(gps.id.unique())

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35, 101, 104, 105, 106,
       107], dtype=int64)

In [7]:
np.sort(gps.date.unique())

array([datetime.date(2014, 1, 6), datetime.date(2014, 1, 7),
       datetime.date(2014, 1, 8), datetime.date(2014, 1, 9),
       datetime.date(2014, 1, 10), datetime.date(2014, 1, 11),
       datetime.date(2014, 1, 12), datetime.date(2014, 1, 13),
       datetime.date(2014, 1, 14), datetime.date(2014, 1, 15),
       datetime.date(2014, 1, 16), datetime.date(2014, 1, 17),
       datetime.date(2014, 1, 18), datetime.date(2014, 1, 19)],
      dtype=object)

There is no car ID assigned with ID 101, 104, 105, 106, and 107. I'll assume that these are trucks ID.

In [8]:
# Total Missing Value
missing_values = gps.isnull().sum()
print('Total Missing Value :\n',missing_values)

# Check the missing value in percentage
missing_values_percentage = (gps.isnull().sum() / len(gps)) * 100
print('\n\nMissing Value Percentage\n',missing_values_percentage)

Total Missing Value :
 Timestamp      0
id             0
lat            0
long           0
date           0
day_of_week    0
weekend        0
hour           0
time           0
dtype: int64


Missing Value Percentage
 Timestamp      0.0
id             0.0
lat            0.0
long           0.0
date           0.0
day_of_week    0.0
weekend        0.0
hour           0.0
time           0.0
dtype: float64


In [9]:
# Merge Data
gps_merged = pd.merge(gps, car, left_on='id', right_on= 'CarID')
gps_merged.drop(['CarID'], axis=1,inplace=True)
gps_merged.head()

Timestamp  id        lat       long        date  day_of_week  \
0 2014-01-06 06:28:01  35  36.076225  24.874689  2014-01-06            0   
1 2014-01-06 06:28:01  35  36.076220  24.874596  2014-01-06            0   
2 2014-01-06 06:28:03  35  36.076211  24.874443  2014-01-06            0   
3 2014-01-06 06:28:05  35  36.076217  24.874253  2014-01-06            0   
4 2014-01-06 06:28:06  35  36.076214  24.874167  2014-01-06            0   

   weekend  hour     time CurrentEmploymentType        CurrentEmploymentTitle  \
0    False     6  Morning             Executive  Environmental Safety Advisor   
1    False     6  Morning             Executive  Environmental Safety Advisor   
2    False     6  Morning             Executive  Environmental Safety Advisor   
3    False     6  Morning             Executive  Environmental Safety Advisor   
4    False     6  Morning             Executive  Environmental Safety Advisor   

            FullName  
0  Willem Vasco-Pais  
1  Willem Vasco-Pais  
2  Willem Vasco-Pais  
3  Willem Vasco-Pais  
4  Willem Vasco-Pais

In [10]:
# Excluding Missing Data
gps_merged.to_csv(r'gps_merged.csv')